Make LPC1768 pinmapping not specific to Re-ARM

tcm0116 opened the pull request:
Currently, the pin mapping for LPC1768-based controllers is assumed to be provided on a per-board basis. The physical pins on the LPC1768 are somewhat arbitrarily mapped to Arduino-style logical pins, which are then referenced wherever necessary. This process makes it quite tedious to define a new LPC1768-based controller.

This PR aims to simplify things by allowing pins files for LPC1768-based controllers to define the pins in terms of the physical pin on the LPC1768. This eliminates the physical-to-logical pin mapping, and makes the LPC1768 HAL more generic. The previous pin mapping utilized a lookup table to convert the logical pin number to the physical pin number. With this change, the physical port and pin number are stored in the logical pin number, and are easily extracted with a few simple bit shifts.

I wanted to get this out there for the team to review. I have run it on a new LPC1768-based controller, and it does work. I'm planning on adding the ability for a HAL to provide encoding and decoding functions for the pin numbers to facilitate using the physical pin numbers in M43  (i.e. M43 P1.6).

Any feedback or testing would be greatly appreciated.

tcm0116 committed:
Make LPC1768 pinmapping not specific to Re-ARM

Roxy-3D commented:
I'm sure there will be suggestions for 'improvement'.   But I am very supportive!

>  I'm planning on adding the ability for a HAL to provide encoding and decoding functions for the pin numbers to facilitate using the physical pin numbers in M43 (i.e. M43 P1.6).

That would be very nice!

Also...  It appears there will be no execution time calculations.  It kind of looks like the pre-processor can do all the bit crunching and make everything small and fast.

p3p commented:
Looking good, removing the obsolete code from trying to transparently support the RAMPS pinout was needed (it wasn't completely arbitrary ;) ), In my implementation I was trying to keep the pin names simple in the pins files (P0_1 ect) rather than use a "long" macro but that's not important, otherwise they're pretty much the same. I was also playing with encoding more information in the pin numbers, (feature support/peripheral number) but we would need to make sure that we could define a pin number bitwidth in the HAL.

tcm0116 commented:
>In my implementation I was trying to keep the pin names simple in the pins files (P0_1 ect) rather than use a "long" macro but that's not important, otherwise they're pretty much the same. I was also playing with encoding more information in the pin numbers, (feature support/peripheral number) but we would need to make sure that we could define a pin number bitwidth in the HAL.

I like that idea. What features were you thinking of including? At a minimum, we could include if the pin is PWM or interrupt capable, which would simplify the corresponding functions to a bit mask.

alexxy commented:
Cool, that is realy needed change =)

tcm0116 commented:
@p3p In the latest commit, I defined the pins per your suggested format and added some feature bits to the pin definition. I also defined `HAL_PIN_TYPE` in the HALs and modified the code to use throughout.

tcm0116 committed:
Add HAL_PIN_TYPE and LPC1768 pin features

p3p commented:
That was fast, I haven't had time to test this but the pins files look much cleaner, and the extra information encoded in the pin number getting rid of the conditional chains is a big improvement, I'l have a better look through later but overall this is exactly what was needed (assuming @thinkyhead  doesn't mind the HAL_PIN_TYPE define).

tcm0116 commented:
It occurred to me that the last commit broke M43 for LPC1768 because you
can't just create pins on the fly since they won't have the feature data. I
think the best solution is to go ahead and create an array of all the P#_#
values and have M43 iterate over that array.

tcm0116 committed:
M43 Updates

tcm0116 commented:
@p3p I spent some time trying to get M43 working, but I'm stuck and I was hoping you could help me. Marlin keeps crashing when I run `M43 W`. I put printouts throughout the 'W' section, but it looks like Marlin crashes at different points with no consistency. I think it happens during the call to `digitalRead`. Any thoughts?

Roxy-3D commented:
> I think the best solution is to go ahead and create an array of all the P#_# values and have M43 iterate over that array.

Does it make sense to only create this array if #define PINS_DEBUGGING is turned on?   On the 32-bit processors memory is at much less of a premium and maybe the array should always be there?

tcm0116 committed:
Move pin map into pinsDebug_LPC1768.h

tcm0116 commented:
>Does it make sense to only create this array if #define PINS_DEBUGGING is turned on?

Good point. I moved it into pinsDebug_LPC1768.h, which only gets included if `PINS_DEBUGGING` is enabled.

thinkyhead commented on Marlin/src/HAL/HAL_AVR/HAL_AVR.h:
```diff
@@ -69,6 +69,8 @@
 #define HAL_TIMER_TYPE uint16_t
 #define HAL_TIMER_TYPE_MAX 0xFFFF
 
+#define HAL_PIN_TYPE int8_t
```
Better if this is a typedef, which can apply to any and all pins:
```cpp
typedef unsigned char pin_t;
```

thinkyhead commented on Marlin/src/HAL/HAL_AVR/HAL_AVR.h:
```diff
@@ -69,6 +69,8 @@
 #define HAL_TIMER_TYPE uint16_t
```
Also…
```cpp
typedef unsigned int timer_t;
```

p3p replied:
I always thought using defines for the types was for style reasons so didn't question it, would have changed it ages ago. The lpc1768 pin constants should probably also be in an enum, and use that as the pin_t type.

thinkyhead replied:
We're definitely now favoring `typedef` and following the `_t` suffix convention. At least, been trying to apply that as I go along…

thinkyhead commented on Marlin/src/HAL/HAL_LPC1768/pinmapping.h:
```diff
@@ -24,13 +24,147 @@
+          (LPC1768_PIN_PORT(p) == 3 && LPC1768_PIN_PIN(p) >= 25 && LPC1768_PIN_PIN(p) <= 26) ? true :
+          (LPC1768_PIN_PORT(p) == 4 && LPC1768_PIN_PIN(p) >= 28 && LPC1768_PIN_PIN(p) <= 29) ? true :
+          false);
+}
```
You can always replace `x = condition ? : true : false` with `x = condition`. So this could be:
```cpp
  return (   (LPC1768_PIN_PORT(p) == 0 && LPC1768_PIN_PIN(p) <= 30)
          || (LPC1768_PIN_PORT(p) == 1 && LPC1768_PIN_PIN(p) <= 31)
          || (LPC1768_PIN_PORT(p) == 2 && LPC1768_PIN_PIN(p) <= 13)
          || (LPC1768_PIN_PORT(p) == 3 && WITHIN(LPC1768_PIN_PIN(p), 25, 26)
          || (LPC1768_PIN_PORT(p) == 4 && WITHIN(LPC1768_PIN_PIN(p), 28, 29)
         );
```
…but since **_Repetition Is Evil_**, this is the best:
```cpp
  constexpr pin_t pin_pin = LPC1768_PIN_PIN(p);
  switch (LPC1768_PIN_PORT(p)) {
    case 0: return pin_pin <= 30;
    case 1: return pin_pin <= 31;
    case 2: return pin_pin <= 13;
    case 3: return WITHIN(pin_pin, 25, 26);
    case 4: return WITHIN(pin_pin, 28, 29);
  }
  return false;
```

tcm0116 replied:
It turns out that the switch statement version won't work. `constexpr` functions on C++11 can only be a single line return statement. Since `VALID_PIN` is used by other `constexpr` functions, then it also has to be `constexpr`.

thinkyhead commented on Marlin/src/HAL/HAL_LPC1768/pinmapping.h:
```diff
@@ -24,13 +24,147 @@
+          p == 6 ? P0_3:
+          p == 7 ? P0_2:
+          -1);
+}
```
```cpp
constexpr pin_t analogInputToDigitalPin(const uint8_t p) {
  constexpr pin_table[] = { P0_23, P0_24, P0_25, P0_26, P1_30, P1_31, P0_3, P0_2 };
  return p < COUNT(pin_table) ? pin_table[p] : -1;
}
```

thinkyhead commented on Marlin/src/HAL/HAL_LPC1768/pinmapping.h:
```diff
@@ -24,13 +24,147 @@
+    return -1;
+
+  return LPC1768_PIN_ADC(p);
+}
```
I ❤️  ternaries.
```cpp
FORCE_INLINE int8_t DIGITAL_PIN_TO_ANALOG_PIN(const pin_t p) {
  return VALID_PIN(p) ? LPC1768_PIN_ADC(p) : -1;
}
```

thinkyhead commented:
Sorry! I could make tweaky code comments all day, but that's obnoxious. Just let me know when this seems reasonable to merge and I'll clean it up at that point.

tcm0116 commented:
@thinkyhead no worries. I enjoy your suggestions as they give me a different way of looking at things. 

I'll incorporate your comments, but I'm struggling with a lockup during M43, but I have a theory. During the `M43 W` command, all non-protected pins are configured as inputs with a pull-up. However, some pins that are outputs are not on the protected list, so their mode is changed but they may still be written to. One examples is the LCD pins, but I'm not certain that the LCD continues to be updated when `M43` is executing. I'll see if I can find a way to check this theory later this weekend when I'm back at my printers.

p3p commented:
@tcm0116 Does it try and change the usb/onboard sd card spi pin modes as well?

tcm0116 commented:
@p3p no. I added those to the sensitive pins list. As I mentioned before,
with print outs at the start of each iteration of the loop, running M43 W
multiple times with no other changes will result in a crash occurring at
different times. That's why I'm wondering if it's a race condition type of
issue.

Roxy-3D commented:
My initial hunch is the SENSITIVE_PINS is going to be slightly different for the ARM processors.

tcm0116 commented:
@Roxy-3D for sure. It currently includes all of the stepper outputs and
heater outputs. I added a HAL_SENSITIVE_PINS define in each HAL and added
the internal SD card and USB pins to it. Before then, the crash was
consistent when trying to manipulate one of those system pins. The
inconsistency of the crash now is what's making me look at other options.

Roxy-3D commented:
> I added a HAL_SENSITIVE_PINS define in each HAL and added the internal SD card and USB pins to it.

Please forgive the ignorant question.   But should the SENSITIVE_PINS be a combination of HAL Sensitive Pins and the Board's Sensitive Pins?    

It would be my guess some of the processor's pins are known to be sensitive and should not be touched accidentally.   But depending upon how a given board is wired up...  Some of the processor pins might go to something 'Sensitive' on one board and not on another board.   So in that mindset...  The HAL for a given processor would have some pins in the list.   But various boards that use that HAL would also add to the list.

tcm0116 commented:
The SENSITIVE_PINS lost was originally a list of known output pins that
shouldn't be reconfigured as part of M43. These included the stepper
control signals and the heater control signals. The reason for this list of
because M43 reconfigures the pins to be inputs with a pull-up, which causes
them to output 5V (or 3.3V). This can be bad as it will enable peripherals
that you don't want enabled. For example, on a new control board I'm
working with, I didn't have the pin for the heater setup properly, and when
I ran M43 W, it caused my heater to turn on and it got to almost 300C
before I realized what was going on.

The HAL_SENSITIVE_PINS list I append to the existing SENSITIVE_PINS list
just gives the HAL the ability to black list pins that shouldn't be
reconfigured by M43. In the case of the LPC1768 HAL, I'm adding the USB
pins and the internal SD card pins. Before adding the USB pins, running M43
W would break the USB by reconfiguring those pins.

I probably just need to try black listing large groups of the pins and
basically do a binary search to determine which one(s) are causing the
lockup. I can then work backwards from there to see why that/those pins are
causing the issue.

On Oct 24, 2017 5:10 PM, "Roxy-3D" <notifications@github.com> wrote:

> I added a HAL_SENSITIVE_PINS define in each HAL and added the internal SD
> card and USB pins to it.
>
> Please forgive the ignorant question. But should the SENSITIVE_PINS be a
> combination of HAL Sensitive Pins and the Board's Sensitive Pins?
>
> It would be my guess some of the processor's pins are known to be
> sensitive and should not be touched accidentally. But depending upon how a
> given board is wired up... Some of the processor pins might go to something
> 'Sensitive' on one board and not on another board.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/8063#issuecomment-339149994>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AByD5WWZrX_B7kpzy9ZjkrZ9T0vZQZR_ks5svmBJgaJpZM4QCOzd>
> .
>

thinkyhead commented:
Perhaps you can tweak `M43`, adding some code so it prints the pin it's testing — then we'll see exactly when the lockup occurs and which pin is most likely the culprit.

In general, `M43` should be considered "slightly dangerous" and is best-used with all extra features disabled, including LCD, SD, etc.

tcm0116 commented:
@thinkyhead as explained above, I've done that. The issue I'm currently
seeing is that it crashes at a different point each time, indicating that
there's some sort of interacting going on. However, I can also try using a
more default config to see if that helps point me in the right direction.

On Oct 24, 2017 6:56 PM, "Scott Lahteine" <notifications@github.com> wrote:

> Perhaps you can tweak M43, adding some code so it prints the pin it's
> testing — then we'll see exactly when the lockup occurs and which pin is
> most likely the culprit.
>
> In general, M43 should be considered "slightly dangerous" and is
> best-used with all extra features disabled, including LCD, SD, etc.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/8063#issuecomment-339171100>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AByD5XOOKteRLnYURoZNmgM3WGhJgZKUks5svnkRgaJpZM4QCOzd>
> .
>

thinkyhead commented:
> as explained above, I've done that. 
> some sort of interacting

Ah, sorry, tl;dr! In that case I would have suggested tossing in a short `safe_delay` after each pin test to give enough time for this interaction to show itself consistently. (Then have it skip that pin on the next test.) I'm just thinking that if we can compile a list of the exact pins then it will give more info than any Google search is likely to provide…

tcm0116 commented:
I was using MYSERIAL.flushTX() after each message, which seemed to do a
good job of ensuring that the message was completely transmitted, but an
additional delay can't hurt.

On Oct 24, 2017 7:21 PM, "Scott Lahteine" <notifications@github.com> wrote:

> as explained above, I've done that.
> some sort of interacting
>
> Ah, sorry, tl;dr! In that case I would have suggested tossing in a short
> safe_delay after each pin test to give enough time for this interaction
> to show itself consistently. (Then have it skip that pin on the next test)
> I'm just thinking that if we can compile a list of the exact pins then it
> will give more info than any Google search is likely to provide…
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/8063#issuecomment-339175055>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AByD5TBUWkBSrKUFDQAZ_EagbRjtUecOks5svn8UgaJpZM4QCOzd>
> .
>

tcm0116 commented:
>Does it make sense to only create this array if #define PINS_DEBUGGING is turned on? 

@Roxy-3D it turns out that the array will be needed for M226, so I'm going to move it back to pinmapping.h.

tcm0116 committed:
Incorporate comments and M226

tcm0116 committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into 2.0.x-lpc1768_pinmapping

tcm0116 committed:
Fix persistent store compilation issues

tcm0116 commented:
I've incorporated @thinkyhead's comments and I may have discovered the cause of `M43` crashing. I had some pins defined in pinmapping.h that are designated as reserved in the LPC1768. I've removed them, so now `M43` shouldn't try to access them. I won't be able to test it until later this week, so it'd be great if someone wouldn't mind giving it a try.

@Spawn32 I updated the MKS SBASE pins file to use the new pin mapping scheme, so could you review it to make sure I didn't mess something up?

tcm0116 committed:
Update pin features

Roxy-3D commented:
>  I had some pins defined in pinmapping.h that are designated as reserved in the LPC1768. I've removed them, so now M43 shouldn't try to access them.

OK.  But if we have a field that describes each pin...   Maybe we should have a 'Reserved' type for pins.  And it should be in the list but M43 is smart enough to ignore 'Reserved' pins?   (I'm not trying to be difficult...  M43 can still access the pin if the person is willing to give the 'Violate the Sensitive Pins List' parameter.)

tcm0116 commented:
The reserved pins are not associated with actual physical pins, so there's
nothing to describe. As such, it doesn't make sense to make them available
for use in the pins files. Trying to use one of them would likely result in
a hardware lockup as I was seeing.

I've made all of the other pins available, except for the ones associated
with the serial port being used (if any), since trying to use one of those
ports for anything would break the ability for the host and the printer to
communicate, which is not very desirable.

Spawn32 commented:
@tcm0116 
Will check and also use the latest for a full test, but coulden't see that all this is merged or am i reading this wrong ?
Are you referring to the version on you git or the one here ? :)

tcm0116 commented:
@Spawn32 this PR isn't merged yet, but you can pull down the branch from my
fork: https://github.com/tcm0116/Marlin/tree/2.0.x-lpc1768_pinmapping

Spawn32 commented:
@tcm0116  You haven't changed anything in the config and _adv.h files ? , used my files from the serial test..

Compiling .pioenvs\LPC1768\src\src\HAL\HAL_STM32F1\HAL_Stm32f1.o
In file included from Marlin\src\HAL\HAL_LPC1768\../../inc/MarlinConfig.h:35:0,
from Marlin\src\HAL\HAL_LPC1768\HardwareSerial.cpp:25:
Marlin\src\HAL\HAL_LPC1768\../../inc/../pins/pins.h:659:8: error: #error "No E stepper plug left for Z2!"
#error "No E stepper plug left for Z2!"
^~~~~
In file included from Marlin\src\HAL\HAL_LPC1768\../../inc/MarlinConfig.h:35:0,
from Marlin\src\HAL\HAL_LPC1768\HAL_timers.cpp:31:
Marlin\src\HAL\HAL_LPC1768\../../inc/../pins/pins.h:659:8: error: #error "No E stepper plug left for Z2!"
#error "No E stepper plug left for Z2!"
^~~~~
In file included from Marlin\src\HAL\HAL_LPC1768\../../inc/MarlinConfig.h:35:0,
from Marlin\src\HAL\HAL_LPC1768\HAL.cpp:23:
Marlin\src\HAL\HAL_LPC1768\../../inc/../pins/pins.h:659:8: error: #error "No E stepper plug left for Z2!"
#error "No E stepper plug left for Z2!"

The good news is that you pin config seems fine, really like the new "way"

tcm0116 commented:
@Spawn32 I didn't change anything other than merging in all the other changes that have been made. It looks like you'll get that error if you've defined more dual stepper drivers than you have extra extruder ports.

In the MKS SBASE pins file, I see you have manually defined the X2 control pins to be the same as the E1 control pins. However, I don't think it's necessary to define the X2 control pins as the pins.h file will automagically define those for you:

https://github.com/MarlinFirmware/Marlin/blob/7b3098bbd5cd45878bce14dedbbf394182859b38/Marlin/src/pins/pins.h#L609-L666

Spawn32 commented:
removed: #define X2_xxx but got the same error, then i just //#define X_DUAL_STEPPER_DRIVERS to get the compiler going and ended up with:

 "pins_MKS_SBASE.h:135:28: error: 'P0_14' was not declared in this scope Compiling .pioenvs\LPC1768\src\src\gcode\control\M3-M5.odefine E_MUX2_PIN         P0_14   // J7-6"

I removed #define E_MUX2_PIN to compile and it compiled without error.

MUX2_PIN isn't really used at the moment but 0_14 should probably be declared in pinmapping.h

My dual x has worked the whole time, no ider what broke it and can't really test the firmware without :/

Spawn32 commented:
I haven't compiled an tested anything fresh off the github since 20.10, are you using a fresh copy here ? something else might have been merged that broke the dual x ?
last thing i compiled that worked with the dual x was you Marlin-2.0.x-hardware_serial and the files there are dated 20.10..  :)

tcm0116 commented:
Hmm. If you don't have `Z_DUAL_STEPPER_DRIVERS` enabled, it shouldn't give you the "No E stepper plug left for Z2!" message. Can you double-check your config?

0.14 is not a valid pin for LPC1768. When I converted your pins file, I just used what was in the old pin mapping file, but apparently there were invalid pins defined. If we can determine what pins would make sense to use for the MK2 multiplexer, I can update the pins file as part of this PR.

Spawn32 commented:
you misunderstood me, i removed :
//#define X2_STEP_PIN        P2_8
//#define X2_DIR_PIN         P2_13
//#define X2_ENABLE_PIN      P4_29
from pins_MKS_SBASE.h and stil got the same error, so then i removed:
#define X_DUAL_STEPPER_DRIVERS from my config and it compiled, but now i am without my second x stepper so can't really test the firmware :/

The pin P0_14 is available on J7 on MKS SBASE so we should be able to use it if we want ? but currently we don't need it for MUX2 since Prusa is only using MUX0-1 for the multiplexer...

tcm0116 commented:
I understand what you did. I don't understand why you're getting an error for Z2 when you don't have `Z_DUAL_STEPPER_DRIVERS` enabled. I'll look at it tonight. 

I'm a bit confused how P0.14 is available. The following section from the manual for the LPC1768 clearly states that pins P0[14:12] are not available. What version of the MKS SBASE do you have? Does it actually have J7? According to the pins file I found (shown below) for the V1.1 and V1.2 doesn't show J7 or P0.14. 

![screenshot_20171025-103647](https://user-images.githubusercontent.com/1868773/32008080-f117e7e2-b970-11e7-938f-170debfc6361.png)

![mks-sbase_pin2](https://user-images.githubusercontent.com/1868773/32008576-25e502a6-b972-11e7-84d6-50672fa93151.jpg)

Spawn32 commented:
never seen that version before, i have a V1.3 but the V1.2 also has the J7, upper right corner:
your picture is also missing the J8... must be a very early version..

File name for this picture is : Mks-Sbase v1.2 Pin Layout, but since i only have a V1.3 here that's the only one i can actually verify :)

![mks-sbase v1 2 pin layout](https://user-images.githubusercontent.com/5589175/32009052-3195900c-b9ae-11e7-9b0b-e7149707b528.jpg)

Spawn32 commented:
thing we are talking passed our self, if i disable Z_DUAL_STEPPER_DRIVERS i dont get the errors, but i really need it since i have 2 z steppers :)

tcm0116 commented:
Does your printer have dual  X steppers, dual Z steppers, or both?

If my changes allow M43 to work, then we can put the P0.14 issue to bed :-)

Spawn32 commented:
OMG !, hahaha, i feel really stupid right now, the confusion is totally on me :)
Can i blame it on me not being able to multitask between Marlin and Altium designer , please ?, lol

i have dual Z steppers and had that enabled always and it worked, but now i notice that in the MKS_SBASE file i had:
#define X2_STEP_PIN        P2_8
#define X2_DIR_PIN         P2_13
#define X2_ENABLE_PIN      P4_29

changed that to: 
#define Z2_STEP_PIN        P2_8
#define Z2_DIR_PIN         P2_13
#define Z2_ENABLE_PIN      P4_29

now it compiles like it should, have no ideer where my brain was when i wrote that in the MKS_Sbase file, and why didn't it throw an error before ? :)

I also have the physical 3 wire link between the Z port and the E1 port so that's probably why it still work even if it was wrong, but doesn't explain why i didn't get an error earlier...

tcm0116 commented:
You shouldn't need to define those Z2 control signals in your pins file.
They should be handled in pins.h automatically.

Spawn32 commented:
That's probably how it should work, but it dosent, if i remove:
#define Z2_STEP_PIN P2_8
#define Z2_DIR_PIN P2_13
#define Z2_ENABLE_PIN P4_29

The same error's comes back :/

I am compiling with :  platformio run -e LPC1768
that's correct isn't it ?

Spawn32 commented:
tested the firmware now, everything works from the MKS TFT display, sd card show up as unn recognized in windows and USB host is off, how did i switch it over to USB hos again ? dont remember :)

But your new pin numbering seems to work perfect :)

tcm0116 commented:
To use the USB host, set SERIAL_PORT in Configuration.h to -1.

Can you try running 'M43 W' to see if it locks up? If not, you can try
grounding those pins on J7 to see what it reports.

Spawn32 commented:
ahh, had forgotten, -1 worked, got usb host back, but my sd card is not recognized and needs formatting according to windows, if i move the card to my pc it ok an no error's on it, any ider what's going on ? works fine in Marlin-bugfix-2.0.x...

Yup, will check those pins with M43..

Spawn32 commented:
Hmm.. how do i "decipher" the pin numbers used by M43 now ?
it reports mye MUX pins as:
READ: PIN: 272            (A 0)  E_MUX1_PIN                         Input  = 1
READ: PIN: 273            (A 0)  E_MUX0_PIN                         Input  = 1
but they are defined as:
#define E_MUX0_PIN         P0_17   // J7-4
#define E_MUX1_PIN         P0_16   // J7-5

tcm0116 commented:
That's something I'm going to fix, so that it will print the LPC1768 pin
number. For now, you have to look at the lower 8 bits in the pin number.
Bits 8-6 are the port and 5-0 are the pin. 272 in binary is 0b100010000, so
bits 8-6 are 0b000 (0) and bits 5-0 are 0b10000 (16), so that's port P0.16.

I'm glad that it didn't crash. Yay.

Spawn32 commented:
tried to ad P0_14 to pinmapping.h but must be more to it, dosen show up in M43 so i can test it, what i did:
inserted:
const pin_t P0_14 = LPC1768_PIN(0, 14, PIN_FEATURE_INTERRUPT);
just before:
const pin_t P0_17 = LPC1768_PIN(0, 17, PIN_FEATURE_INTERRUPT);

and added it to this line:
P0_14,P0_17, P0_18, P0_19, P0_20, P0_21, P0_22, P0_23, P0_24,
and enabled it in my MKS_SBASE pin config..

btw, any ider why the sd card communication is down ? hard work to insert it in my pc to re write it, lol

tcm0116 commented:
Without modifying the pinmapping.h file, it prints the READ lines for pins 272 (P0.16) and 273 (P0.17) when you ground them on J7, but nothing when you ground the P0.16 pin on J7?

You're missing one change to add P0.14. You need to change the following line in the VALID_PIN function:
```
(LPC1768_PIN_PORT(p) == 0 && WITHIN(LPC1768_PIN_PIN(p), 15, 30))  ||
```
to:
```
(LPC1768_PIN_PORT(p) == 0 && WITHIN(LPC1768_PIN_PIN(p), 14, 30))  ||
```

I'm not really sure why the SD card would stop working. Maybe @p3p can provide some insight as I'm not overly familiar with how that works.

Spawn32 commented:
I haven't grounded 16 and 17 yet, will do them now together with 14 :)

Spawn32 commented:
Before:
READ: PIN: 272            (A 0)  E_MUX1_PIN                         Input  = 1
READ: PIN: 270            (A 0)  E_MUX2_PIN                         Input  = 0
READ: PIN: 273            (A 0)  E_MUX0_PIN                         Input  = 1

Then i set 272 and 273 to gnd and 270 to 5v with jumpers, did a "plain" M43 and none of them changed.
So i did a M43 P272 and got a full crash :/

Spawn32 commented:
actually had to power the board down, reset wasn't enough...

tcm0116 commented:
270 is P0.14. That's so strange since the manual says that P0.14 is not available. @p3p any ideas why that might be?

272 isn't the right parameter to pass to `M43 P`. I'm working on cleaning that up, but it shouldn't have crashed, so I'll look at that as well.

tcm0116 commented:
@Spawn32 try running `M43 W` and then connecting the pins to ground. Since your pin 270 actually reported as having a value of 0, I'm thinking it didn't actually work.

Spawn32 commented:
did a new P272, this time it did something before crashing, log included :)
[M43Log.txt](https://github.com/MarlinFirmware/Marlin/files/1416344/M43Log.txt)

One M43 W coming up in a few seconds :)

Spawn32 commented:
Hmm...
It seems to read 0.16 and 0.17 just fine:
READ: PIN: 273            (A 0)  E_MUX0_PIN                         
READ: .                          MISO_PIN                           
READ: PIN:  47                   REF_CLK                            
READ: PIN:  47                   REF_CLK                            
READ: PIN: 272            (A 0)  E_MUX1_PIN                         
READ: PIN:  47                   REF_CLK                            
READ: PIN:  47                   REF_CLK                            

But look at 0.14, strange, probably shoulden't use that one for anything, makes me wonder why MKS made it available on J7:

READ: PIN: 271            (A 0)  SCK_PIN                            
READ: PIN:  47                   REF_CLK                            
READ: PIN:  47                   REF_CLK                            
READ: PIN:  47                   REF_CLK               

it should at least have said 270 :)

p3p commented:
P0.14 according to the LPC176x specification is reserved and can't be used for anything, like you said, I don't understand why its been used .. 

@Spawn32 Have you tried formatting the sd card, windows may be more lenient with FAT corruption.

tcm0116 commented:
271 is P0.15. I think there's a typo on that MKS SBASE pins file.

I have a change ready to test that prints the actual pin number (such as
P0.15) instead of 271. I'm going to look at how to get the P parameter of
M43 to accept something like 0.15 later tonight.

Spawn32 commented:
@p3p  sd card works fine if i put back my Marlin-bugfix-2.0.x binary, and have scanned it with windows for errors ...

Spawn32 commented:
@tcm0116 
Well, if it's a typo it becuse of MKS's pin description on that picture i gave you, will try to change it to 0.15 and see what happens.

You want me to download a new version from you github, it's ready ?

tcm0116 commented:
I haven't pushed any changes yet. I'll will in a few hours.

p3p commented:
I'm unsure the only thing I can think that would stop the sd card working is if the pins are reconfigured after the low level driver is started, the usb ect doesn't use the HAL functions.

Spawn32 commented:
@p3p will download a more resent Marlin-bugfix-2.0.x, the one i used is 4-5 days old, maybe a merge broke it, will test...

Spawn32 commented:
@tcm0116 
Hmm... thinking about it, from the M43 W log it looks like it was triggering both 0.14 and 0.15 when i toggled it, might look like they are connected some how in the MCU and shoulden't be touched ?

tcm0116 commented:
Since 0.14 is reserved, there's no telling what the MCU writes to those
bits in the status resister. If you're getting a report for 0.15 when you
ground that pin, then that's the one.

Spawn32 commented:
@p3p 
Can't compile Marlin-bugfix-2.0.x anymore, did something change in the config or the _adv the last days ? i am using my old once.. see log:
[compile.txt](https://github.com/MarlinFirmware/Marlin/files/1416445/compile.txt)

tcm0116 commented:
I had that same issue. You can incorporate this commit: https://github.com/MarlinFirmware/Marlin/commit/1b6497a3a630cd93fdad268b1b196a925047a005

p3p commented:
aah @thinkyhead being a little overzealous on the cleanup, its best to use the types defined by the library when using a libs api.

Spawn32 commented:
@tcm0116 are those 3 files up to date in your Marlin-2.0.x-lpc1768_pinmapping ?
so i can just copy them ?

Spawn32 commented:
@tcm0116 
MKS's pinout diagram is wrong, removed the P0_14 from pinmaping.h and edited MUX2 from 14 to 15 and now it's correct:

READ: PIN: 271            (A 0)  E_MUX2_PIN                         
READ: .                          SCK_PIN                            
READ: PIN:  47                   REF_CLK                            
READ: PIN:  47                   REF_CLK                            
READ: PIN:  47                   REF_CLK                            
READ: PIN:  47                   REF_CLK

Spawn32 commented:
hehe, was not that easy, will just have to wait until someone merges https://github.com/MarlinFirmware/Marlin/commit/1b6497a3a630cd93fdad268b1b196a925047a005
it's broken right now, so we need that merged @p3p 😃

Spawn32 commented:
@tcm0116 for the record, i am still new to this github things, is there  easy way of "incorporate this commit:" when it's not merged yet, without manually editing the files :)

tcm0116 commented:
@Spawn32 you can try running 'git
cherry-pick 1b6497a3a630cd93fdad268b1b196a925047a005' from the command
line.

tcm0116 committed:
Update MKS SBASE pins

tcm0116 committed:
Use native LPC1768 pin numbers in M42, M43, and M226

tcm0116 committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into 2.0.x-lpc1768_pinmapping

tcm0116 commented:
@Spawn32 I just pushed a few commits. The first updates the MKS SBASE pins file. The second hopefully adds the ability to print the native pin number in M43 and also accept the native pin numbers in M43, M43, and M226. If you pull down my branch, you should be able to do things like `M43 P0.15` to see the status of P0.15. I haven't tested it since I'm away from my printer, so hopefully it'll work :-)

Spawn32 commented:
@tcm0116 
Works perfect now, tested M43, M43 W , M43 P0.15 and M43 P0.15 W no crashes and it seems correct, good job :)

Think you should add:
#define Z2_STEP_PIN        P2_8
#define Z2_DIR_PIN         P2_13
#define Z2_ENABLE_PIN      P4_29
Permanently to the MKS_SBASE pin file, you get the compile error without it...

and any chanse we can get https://github.com/MarlinFirmware/Marlin/commit/1b6497a3a630cd93fdad268b1b196a925047a005 merged today ? the MKS guy's are on me and want's to test 2.0 :)

tcm0116 commented:
@Spawn32 that's great news! I'm glad that it's working considering I couldn't test the changes. 

I removed the Z2 control pins and verified there were no compilation issues with `Z_DUAL_STEPPER_DRIVERS` with them removed. I wonder if there's something else causing that error. Can you upload the config files you were using with my branch?

Spawn32 commented:
Shoulden't
 be anything special in them but i might have overlooked something :)
here they are:
[Marlin.zip](https://github.com/MarlinFirmware/Marlin/files/1419470/Marlin.zip)

Roxy-3D commented:
Let's cross our fingers and hope things are better after the merge!

fiveangle commented:
> define the pins in terms of the physical pin on the LPC1768. This eliminates the physical-to-logical pin mapping

How support for every MCU should have been implemented from the start ! :raised_hands: 

-=dave

Spawn32 commented:
@tcm0116  have you had time to check my config's ?
if they look normal i will start back tracking the commits until 20.10 as @Roxy-3D suggested to see when the sd-card problem started...

Just have to finish some pcb layout i am doing, a real busy bee this day's :D

tcm0116 commented:
@Spawn32 #8130 was just merged and should fix your dual stepper issue.

Spawn32 commented:
Compiles with out any warring, works like it should, no dual z bitching, nice @tcm0116 
so only thing left for me is the bloody sd card problem, must start backtracking merges today :/

Spawn32 commented:
oh well, it's something in my config's, used a plain copy of github and just edit'ed the config to set the correct board and now i have sd card again, strange :/

Did you go over the config's i attached @tcm0116 , would have been nice to know what broke it..

tcm0116 commented:
I looked at them, but was focused on the dual stepper issue. They were a
bit out of date, so you might consider using a compare tool to compare them
to the default configuration files to see what needs to be updated.

Spawn32 commented:
yes, seems like i have to do that :)

So what do you think so fare ? 
Has it own Pi Zero / Octoprint port, rgb (3Amp), removable power modules in case you toast them, etc, etc and 100% faster than a STM M4, it's a cortex A7 @216Mhz, twice the cpu power , lol :)

<img width="1625" alt="pre_02" src="https://user-images.githubusercontent.com/5589175/32144266-90b4262e-bcb6-11e7-9754-685c8c23e6e9.png">

The J16 is for Power loss, can use a MKS powerloss module or like the new Prusas MK3...

What do you think @tcm0116  @Roxy-3D  @p3p @thinkyhead  @AnHardt ? :) negative comment's will be ignored, lol

BTW, can you guy's start working on trippel hoste mode and maybe FreeRTOS ? 😈 

Haven't looked to much in to this host thing yet, but couldn't you send "busy" to all connected hosts and only accept emergency commands if a another host is being used and only be allowed to switch if the other host is finished ? and let all host's get transmits ? (i probably didn't think this through enough)...

apballard commented:
Hi All

I've been having problems with my re-arm for a few days now.  I've narrowed it down to:

commit 162ef7538c944e09d25fa226a19e3a81f96b8897

My SD Card no longer reads on my Mac when connected using usb.
When connected to my octopi I get serial disconnects, slowed down lcd refreshes, the whole re-arm slows down, no response to clicks.  G29 Staggers, taking about 10 seconds per probe.

Its only when I have the octopi connected, however this was not a problem before this.

Any suggestions?

apballard commented:
The printer will have bursts of normal operation for a few seconds, to a couple minutes, but then go back to being slow and jerky.

Reverting this commit allows my re-arm and printer to work fine.

https://github.com/apballard/BigBoxMarlin/tree/Marlin2.0_BigBox for config

Spawn32 commented:
@apballard 
Do you have https://github.com/MarlinFirmware/Marlin/commit/50b4f9cbf7ba1400c0207a2fdd1bd70a8df8fda5 in your git ?

apballard commented:
@Spawn32 I do have it.

My last tests today included everything up to today.

Spawn32 commented:
Then i have no idea, one of the expert's here will probably soon help you :)

tcm0116 commented:
@apballard I was just looking through the repository link you provided so I can try to replicate your issue, but I don't see any unique configuration files for me to pull down. Am I missing something?

tcm0116 commented:
@apballard My current guess is that there's some configuration option that's causing the issues you're seeing. I'm currently running a G33 calibration on my delta machine with a Re-ARM and the latest bugfix-2.0.x, and it's running very smoothly. I even tried copying a file to the onboard SD card during the calibration, but that didn't affect the operation of the printer.

If you have some time and are willing to help, it would be great if you could start reverting the changes to your configuration the default values to see if you can isolate which one is causing the issue. I was planning on trying that myself while I had a few spare minutes this afternoon, but I couldn't seem to locate your configuration files.

p3p commented:
@tcm0116 I've been going through this PR (a bit late now :wink:), I set the pin constants to be an enum to get the type safety (curse those macros stopping us doing things properly) and traced these required changes, there could be more but I'm realy struggling for time atm and dont know if when I can get back to it so I'l point you to this commit rather than make a PR.
https://github.com/p3p/Marlin/commit/e9f21e7c00fadae5b00c0c85ea0bb608a6efda63

tcm0116 commented:
@p3p did you see the changes I made in #8119? I'll run with your commit.
Thanks!

On Oct 29, 2017 4:54 PM, "Chris Pepper" <notifications@github.com> wrote:

> @tcm0116 <https://github.com/tcm0116> I've been going through this PR (a
> bit late now 😉), I set the pin constants to be an enum to get the type
> safety (curse those macros stopping us doing things properly) and traced
> these required changes, there could be more but I'm realy struggling for
> time atm and dont know if when I can get back to it so I'l point you to
> this commit rather than make a PR.
> p3p/Marlin@e9f21e7
> <https://github.com/p3p/Marlin/commit/e9f21e7c00fadae5b00c0c85ea0bb608a6efda63>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/8063#issuecomment-340304010>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AByD5UhQmgY_K2d-Zd8yottlxMEYQpGHks5sxPQagaJpZM4QCOzd>
> .
>

tcm0116 commented:
@apballard I've incorporated @p3p's changes into a branch in my repository. Would you mind pulling it down and giving it a try with your configuration to see if it makes any difference?

https://github.com/tcm0116/Marlin/tree/2.0.x-pin_mapping

apballard commented:
@tcm0116 
Thanks for the suggestions.  Let me try your repo.  I'll also start cutting down on my config. I do have quite a few features enabled.
It only happens when connected to a usb device.  If I connect to my raspberry pi, it causes problems.

tcm0116 commented:
@apballard sounds like a plan. Thanks for helping out with this. It's
difficult to diagnose on my end since I can't replicate it. I do have a
Raspberry Pi, so I'll give that a try tonight to see if that does the
trick.

apballard commented:
Ok, so I've cut down to bare minimum config. still no go.  It only has issues when connected to raspberry pi.  I've run out of time today, but I think next step is I'll re-install my raspberry pi.

When connected to my computer, it has no issues, until the sd card is mounted and you try access it.

This was not an issue before, so I'm guessing its got to do with the SD card and accessing via the usb.

I have a usb cable that has two usb female ends.  one is for power only, other is for power and data.  If I plug power only into raspberry, its fine, and if I plug in the data side as well, it does not work. thats why I'm guessing usb.

tcm0116 commented:
What OS are you running on your computer?

tcm0116 commented:
@apballard I just did some testing with my delta machine running a Re-ARM board with a RPi3 connected running OctoPrint. I did several `G29` commands, but didn't see any degradation in performance of the printer. I'm definitely lost with this.

apballard commented:
@tcm0116 I’m not sure if delta vs Cartesian would make a diff. Let me fiddle around a bit more. Might only get to it later this week. Seems like I’m the only one, so I’ll give it more effort.

Both windows and Mac. Neither read the contents any more after the update. It could be an electronic issue, just aggravated by the update, so I’m gonna strip down the board to only power and usb and work my way back.

apballard commented:
@tcm0116 Thanks for your help in the past.
I finally got time to investigate why my re-arm was freezing.
I found that from my Rumba config, I had left the option E0_AUTO_FAN_PIN set to pin 8.  When I tested a minimal config, I would have viewed this as a necessary minimum, but you can see my stupidity, since I'm not using this function on the ramps.

The pin re-config must have actually "fixed" something to make it not work.  I'm getting a fan extender for my ramps board, and I'll map the correct pin in time.

Thanks again.

tcm0116 commented:
@apballard glad you got it figured out!
